### Natural Language Processing - Homework 2

Submitted by: Apurv Mittal

In [1]:
# Importing necessary libraries
import nltk
from nltk import word_tokenize
from urllib import request
import matplotlib as plt
import os
from nltk.tokenize import RegexpTokenizer
import numpy as np
from sklearn.preprocessing import minmax_scale
from IPython.display import Image
from IPython.core.display import HTML 
import requests
import urllib.request
import time
from bs4 import BeautifulSoup
import re
from nltk.corpus import stopwords
from nltk.tokenize import word_tokenize
from string import punctuation
#from nltk.book import *
import warnings

In [2]:
#Ignore Warnings

warnings.filterwarnings('ignore')

Reading text files from http://www.gutenberg.org/ebooks/bookshelf/215 based on grade levels.

We will take grade 2 through grade 6 texts for our analysis.

#### Reference to read text files: 
https://www.pythontutorial.net/python-basics/python-read-text-file/

http://www.nltk.org/book/ch03.html


#### Read Grade 2 Book

In [3]:
url = "https://www.gutenberg.org/cache/epub/14668/pg14668.txt"
response = request.urlopen(url)
grade2book = response.read().decode('utf8')
tokens2 = word_tokenize(grade2book)
text2 = nltk.Text(tokens2)
print(text2)

<Text: The Project Gutenberg EBook of McGuffey 's Second...>


#### Read Grade 3 Book

In [4]:
url = "https://www.gutenberg.org/cache/epub/14766/pg14766.txt"
response = request.urlopen(url)
grade3book = response.read().decode('utf8')
tokens3 = word_tokenize(grade3book)
text3 = nltk.Text(tokens3)
print(text3)

<Text: The Project Gutenberg EBook of McGuffey 's Third...>


#### Read Grade 4 Book

In [5]:
url = "https://www.gutenberg.org/cache/epub/14880/pg14880.txt"
response = request.urlopen(url)
grade4book = response.read().decode('utf8')
tokens4 = word_tokenize(grade4book)
text4 = nltk.Text(tokens4)
print(text4)

<Text: The Project Gutenberg EBook of McGuffey 's Fourth...>


#### Read Grade 5 Book

In [6]:
url = "https://www.gutenberg.org/cache/epub/15040/pg15040.txt"
response = request.urlopen(url)
grade5book = response.read().decode('utf8')
tokens5 = word_tokenize(grade5book)
text5 = nltk.Text(tokens5)
print(text5)

<Text: The Project Gutenberg EBook of McGuffey 's Fifth...>


#### Read Grade 6 Book

In [7]:
url = "https://www.gutenberg.org/cache/epub/16751/pg16751.txt"
response = request.urlopen(url)
grade6book = response.read().decode('utf8')
tokens6 = word_tokenize(grade6book)
text6 = nltk.Text(tokens6)
print(text6)

<Text: The Project Gutenberg EBook of McGuffey 's Sixth...>


#### Method to get the Vocab Size and normalize the score

##### Reference: Homework Example shared with the class

In [8]:
#### Method to get the Vocab Size and normalize the score

def n_vocab_size(*arg):
    vocab_size = np.array([])
    vocab_size_norm = np.array([])
    
    #### Getting the Vocab Size
    for text in arg:
        vocab_size = np.append(vocab_size,len(set(text)))
    
    #### Normalizing using the formula 
    for vsize in vocab_size:
        vocab_size_norm = np.append(vocab_size_norm,(vsize - vocab_size.min()) /
                                                    (vocab_size.max() - vocab_size.min()))
    
    #### Normalizing using sklearn preprocessing 
    vocab_size_norm_sklearn = minmax_scale(vocab_size, feature_range=(0,1), axis=0)
    
    return(vocab_size,vocab_size_norm,vocab_size_norm_sklearn)

The above method is calculating the normalized score from an array of scores of different texts. There are two methods covered. `1` method is the manual calculation by using minimum vocabulary size of a text and comparing with maximum vocabulary size. `2` method is using an in-build function from sklearn. Its also doing the same calculations.

For rest of this homework, we will be using SKlearn method for normalize calculations.

In [9]:
vocab_size = n_vocab_size(text2,text3,text4,text5,text6)

Invoking the function to get the normalized vocabulary size from the function defined above.

In [10]:
type(vocab_size)

tuple

In [11]:
norm_vocab_list = vocab_size[2].tolist()

Converting the tuple to a list for easy interpretation and calculations.

In [12]:
print("Normalized vocabulary size for Grade 2 Book: ", norm_vocab_list[0])
print("Normalized vocabulary size for Grade 3 Book: ", norm_vocab_list[1])
print("Normalized vocabulary size for Grade 4 Book: ", norm_vocab_list[2])
print("Normalized vocabulary size for Grade 5 Book: ", norm_vocab_list[3])
print("Normalized vocabulary size for Grade 6 Book: ", norm_vocab_list[4])

Normalized vocabulary size for Grade 2 Book:  0.0
Normalized vocabulary size for Grade 3 Book:  0.052202160610410164
Normalized vocabulary size for Grade 4 Book:  0.48032031427060506
Normalized vocabulary size for Grade 5 Book:  0.7757044647578756
Normalized vocabulary size for Grade 6 Book:  0.9999999999999999


This is the normalized vocabulary size by the grades. As expected the normalized score is lowest for Grade 2 and its the highest for Grade 6. As the grades increase the text and vocabulary increases.

#### Long Words

Now lets start with identifying the long words. For convenience and to avoid doing the same calculations multiple times, we have defined a function to identify the long words.

For this exercise we have defined words longer than 10 to be long words.

#### Reference for the function below: https://www.nltk.org/book/ch01.html

In [13]:
def longwords(words):
    v=set(words)
    long = [w for w in v if len(v)>10]
    return (long)

In [14]:
long2 = longwords(text2)
long3 = longwords(text3)
long4 = longwords(text4)
long5 = longwords(text5)
long6 = longwords(text6)

Passing the gradewise text to the `longwords` function to get the list of long words. Now, we will pass these words to the normalized vocabulary score calculation function defined earlier to get the score.

In [15]:
vocab_size = n_vocab_size(long2,long3,long4,long5,long6)

In [16]:
norm_long_vocab_list = vocab_size[2].tolist()

Converting from tuples to list.

In [17]:
print("Normalized Long Words vocabulary size for Grade 2 Book: ", norm_long_vocab_list[0])
print("Normalized Long Words vocabulary size for Grade 3 Book: ", norm_long_vocab_list[1])
print("Normalized Long Words vocabulary size for Grade 4 Book: ", norm_long_vocab_list[2])
print("Normalized Long Words vocabulary size for Grade 5 Book: ", norm_long_vocab_list[3])
print("Normalized Long Words vocabulary size for Grade 6 Book: ", norm_long_vocab_list[4])

Normalized Long Words vocabulary size for Grade 2 Book:  0.0
Normalized Long Words vocabulary size for Grade 3 Book:  0.052202160610410164
Normalized Long Words vocabulary size for Grade 4 Book:  0.48032031427060506
Normalized Long Words vocabulary size for Grade 5 Book:  0.7757044647578756
Normalized Long Words vocabulary size for Grade 6 Book:  0.9999999999999999


As expected the normalized score increses with grade level. Its lowest for Grade 2 and the highest for Grade 6 as many more longer words are added in grade 6 books compared to Grade 2.

#### Text Difficulty Score

Defining a function to calculate Lexical Diversity with reference from the text book chapter 1.

https://www.nltk.org/book/ch01.html

In [18]:
def lexical_diversity(text):
    return len(set(text)) / len(text)

In [19]:
print("Lexical Diversity Score of Grade 2 Book: ",lexical_diversity(text2))
print("Lexical Diversity Score of Grade 3 Book: ",lexical_diversity(text3))
print("Lexical Diversity Score of Grade 4 Book: ",lexical_diversity(text4))
print("Lexical Diversity Score of Grade 5 Book: ",lexical_diversity(text5))
print("Lexical Diversity Score of Grade 6 Book: ",lexical_diversity(text6))

Lexical Diversity Score of Grade 2 Book:  0.1588342153068478
Lexical Diversity Score of Grade 3 Book:  0.12404916695180691
Lexical Diversity Score of Grade 4 Book:  0.12347441296124474
Lexical Diversity Score of Grade 5 Book:  0.1128618252181811
Lexical Diversity Score of Grade 6 Book:  0.10087260442906655


The Lexical Diversity score is falling as we go higher in the grades. For example Grade 2 Lexical Diversity is 15.88% while that of Grade 6 book has reduced to 10.08%. This is surprising as one would expect the Lexical Diversity to be higher for higher grade books. The number of unique words should be lot more in the higher grade books, thus leading to higher Lexical Diversity. This needs to be analyzed further as its against our expectations.

To calculate `Text Difficulty Score`, we will take a mean of `Lexical Diversity Score`, `Normalized Vocabulary Score` and `Normalized Long Vocabulary Score`.

In [20]:
text_difficulty2 = (lexical_diversity(text2)+norm_long_vocab_list[0]+norm_vocab_list[0])/3
text_difficulty3 = (lexical_diversity(text3)+norm_long_vocab_list[1]+norm_vocab_list[1])/3
text_difficulty4 = (lexical_diversity(text4)+norm_long_vocab_list[2]+norm_vocab_list[2])/3
text_difficulty5 = (lexical_diversity(text5)+norm_long_vocab_list[3]+norm_vocab_list[3])/3
text_difficulty6 = (lexical_diversity(text6)+norm_long_vocab_list[4]+norm_vocab_list[4])/3

In [21]:
print("Text Difficulty Score of Grade 2 Book: ",text_difficulty2)
print("Text Difficulty Score of Grade 3 Book: ",text_difficulty3)
print("Text Difficulty Score of Grade 4 Book: ",text_difficulty4)
print("Text Difficulty Score of Grade 5 Book: ",text_difficulty5)
print("Text Difficulty Score of Grade 6 Book: ",text_difficulty6)

Text Difficulty Score of Grade 2 Book:  0.05294473843561593
Text Difficulty Score of Grade 3 Book:  0.07615116272420908
Text Difficulty Score of Grade 4 Book:  0.3613716805008183
Text Difficulty Score of Grade 5 Book:  0.5547569182446441
Text Difficulty Score of Grade 6 Book:  0.7002908681430221


Above, a `Text Difficutly Score` is printed for all grade books. As we expect the text difficulty increases with grades. Grade 2 sees the lowest `Text Difficutly Score` at `0.05` while it gradually increases by the grades and highest for Grade 6 at `0.70`.

#### - End of Homework 2 - 